official/vision/beta/projects/yolo/optimization/sgd_torch.py

```diff
@@ -0,0 +1,339 @@
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import array_ops
+from tensorflow.keras.optimizers.schedules import LearningRateSchedule
+from tensorflow.python.training import gen_training_ops
+
+import tensorflow as tf
+import re
+import logging
+
+try:
+  from keras.optimizer_v2.optimizer_v2 import _var_key
+except:
+  def _var_key(var):
+    """Key for representing a primary variable, for looking up slots.
+    In graph mode the name is derived from the var shared name.
+    In eager mode the name is derived from the var unique id.
+    If distribution strategy exists, get the primary variable first.
+    Args:
+      var: the variable.
+    Returns:
+      the unique name of the variable.
+    """
+
+    # pylint: disable=protected-access
+    # Get the distributed variable if it exists.
+    if hasattr(var, "_distributed_container"):
+      var = var._distributed_container()
+    if var._in_graph_mode:
+      return var._shared_name
+    return var._unique_id
+
+
+class SGDTorch(tf.keras.optimizers.Optimizer):
+  """Optimizer that simulates the SGD module used in pytorch. 
+  
+  
+  For details on the differences between the original SGD implemention and the 
+  one in pytorch: https://pytorch.org/docs/stable/generated/torch.optim.SGD.html.
+  This optimizer also allow for the usage of a momentum warmup along side a 
+  learning rate warm up, though using this is not required. 
+
+  Example of usage for training:
+  ```python
+  opt = SGDTorch(learning_rate, weight_decay = 0.0001)
+  l2_regularization = None
+
+  # iterate all model.trainable_variables and split the variables by key 
+  # into the weights, biases, and others.
+  optimizer.search_and_set_variable_groups(model.trainable_variables)
+
+  # if the learning rate schedule on the biases are different. if lr is not set 
+  # the default schedule used for weights will be used on the biases. 
+  opt.set_bias_lr(<lr schedule>)
+
+  # if the learning rate schedule on the others are different. if lr is not set 
+  # the default schedule used for weights will be used on the biases. 
+  opt.set_other_lr(<lr schedule>)
+  ```
+  """
+
+  _HAS_AGGREGATE_GRAD = True
+
+  def __init__(self,
+               weight_decay=0.0,
+               learning_rate=0.01,
+               momentum=0.0,
+               momentum_start=0.0,
+               warmup_steps=1000,
+               nesterov=False,
+               sim_torch=False,
+               name="SGD",
+               weight_keys=["kernel", "weight"],
```

Comment:
We can have kernel|weight using regex, right? If so, this can just be a string.

Reply:
making this a string may miss some other edge cases for other models. the other optimizers use List[str] as well.

Reply:
done